### PR TITLE
[WIP] Add LXMERT for visual question answering

### PR DESCRIPTION
**Pull request by gejinchen**
1. Still need to refactor the datasets and change the code accordingly.
2. Convergence runs on whole datasets are yet to be done.

### CODE REVIEW

File: examples/question-answering/README.md [hunk 1]
```diff
@@ -56,3 +56,31 @@ f1 = 88.52
...
 -->
+
+# Visual question answering
+
+`run_vqa.py` allows you to fine-tune the LXMERT model for visual question answering datasets such as VQA v2 and GQA, which are available on our [hub](https://huggingface.co/datasets).
+
+## Fine-tuning LXMERT on GQA
+
+```bash
+python examples/question-answering/run_vqa.py \
```
Comment (michaelbenayoun) [inline examples/question-answering/README.md:67]: Please add the pod type.

File: examples/question-answering/run_vqa.py [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+    HfArgumentParser,
+    default_data_collator,
+    set_seed,
+)
+from transformers.trainer_utils import get_last_checkpoint
+from transformers.utils import check_min_version
+from transformers.utils.versions import require_version
+
+
+# from utils_qa import postprocess_qa_predictions
```
Comment (michaelbenayoun) [inline examples/question-answering/run_vqa.py:53]: Why is this commented?

File: examples/question-answering/run_vqa.py [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+import datasets
+import numpy as np
+from datasets import load_dataset, load_metric
+
+import transformers
+from optimum.graphcore import IPUConfig, IPUTrainer
+from optimum.graphcore import IPUTrainingArguments as TrainingArguments
+from optimum.graphcore.data import pad_on_batch_axis
+
+# from trainer_qa import QuestionAnsweringTrainer
```
Comment (michaelbenayoun) [inline examples/question-answering/run_vqa.py:37]: Are you sure you dont want to use the `QuestionAnsweringTrainer`?

File: examples/question-answering/run_vqa.py [hunk 4]
```diff
@@ -0,0 +1,532 @@
+#!/usr/bin/env python
```
Comment (michaelbenayoun) [inline examples/question-answering/run_vqa.py:1]: You will need to add the reference diff file to get tests to pass, I can help you with that.

File: examples/question-answering/run_vqa.py [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+    )
+
+    log_level = training_args.get_process_log_level()
+    logger.setLevel(log_level)
+    datasets.utils.logging.set_verbosity(log_level)
+    transformers.utils.logging.set_verbosity(log_level)
+    transformers.utils.logging.enable_default_handler()
+    transformers.utils.logging.enable_explicit_format()
+
+    # Log on each process the small summary:
```
Comment (michaelbenayoun) [inline examples/question-answering/run_vqa.py:241]: You can remove those comments.

File: examples/question-answering/run_vqa.py [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+    # Data collator
+    # We have already padded to max length if the corresponding flag is True, otherwise we need to pad in the data
+    # collator.
+    data_collator = (
+        default_data_collator
+        if data_args.pad_to_max_length
+        else DataCollatorWithPadding(tokenizer, pad_to_multiple_of=8 if training_args.fp16 else None)
+    )
+
+    # if training_args.do_train and training_args.pad_on_batch_axis:
```
Comment (michaelbenayoun) [inline examples/question-answering/run_vqa.py:425]: Why is this commented? Should we remove this?

File: optimum/graphcore/models/lxmert/modeling_lxmert.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        logger.info("Image embedding         --> IPU 0")
+        self.lxmert.encoder.visn_fc = poptorch.BeginBlock(self.lxmert.encoder.visn_fc, "Image embedding", ipu_id=0)
+
+        # Language layers
+        for index, layer in enumerate(self.lxmert.encoder.layer):
+            if self.config.recompute_checkpoint_every_layer:
+                h = recomputation_checkpoint(layer)
+                self._hooks.append(h)
+            self.lxmert.encoder.layer[index] = poptorch.BeginBlock(layer, f"Language layer{index}", ipu_id=1)
+            logger.info(f"Language layer {index:<2}       --> IPU 1")
```
Comment (michaelbenayoun) [inline optimum/graphcore/models/lxmert/modeling_lxmert.py:55]: Why are ipu_ids hardcoded?

File: optimum/graphcore/models/lxmert/modeling_lxmert.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        self,
+        input_ids,
+        visual_feats,
+        visual_pos,
+        attention_mask=None,
+        token_type_ids=None,
+        labels=None,
+        visual_attention_mask=None,
+    ):
+        lxmert_output = self.lxmert(
```
Comment (michaelbenayoun) [inline optimum/graphcore/models/lxmert/modeling_lxmert.py:91]: Are you changing the forward pass implementation or can we use the `super()`  method?

Comment (gejinchen) [reply optimum/graphcore/models/lxmert/modeling_lxmert.py:91]: Yes, the forward() is changed. The only difference is that now the forward() could support soft labels from the VQA v2 dataset.

File: examples/question-answering/run_vqa.py [hunk 9]
```diff
@@ -0,0 +1,532 @@
...
+
+    if training_args.push_to_hub:
+        trainer.push_to_hub(**kwargs)
+    else:
+        trainer.create_model_card(**kwargs)
+
+
+def _mp_fn(index):
+    # For xla_spawn (TPUs)
+    main()
```
Comment (jimypbr) [inline examples/question-answering/run_vqa.py:528]: remove this

File: examples/question-answering/run_vqa.py (hunk 3 above)
Comment (gejinchen) [reply examples/question-answering/run_vqa.py:37]: I am pretty sure. VQA and SQuAD are actually quite different tasks. VQA's prediction head just outputs a single prediction, so it's more like image classification.

File: examples/question-answering/run_vqa.py (hunk 2 above)
Comment (gejinchen) [reply examples/question-answering/run_vqa.py:53]: It should be deleted. VQA's outputs are not substrings of contexts.

File: examples/question-answering/run_vqa.py (hunk 6 above)
Comment (gejinchen) [reply examples/question-answering/run_vqa.py]: It turned out to be quite complicated to handle the dataloader remainder for soft labels. Considering that most tasks just do drop_last, I think we should just do the same for VQA for now.

File: optimum/graphcore/models/lxmert/modeling_lxmert.py (hunk 7 above)
Comment (gejinchen) [reply optimum/graphcore/models/lxmert/modeling_lxmert.py:55]: There is only LXMERT-base. No small or medium ... One could of course override and add more layers to each component, and since the components are not the same in memory consumption, one of the IPUs might go OOM earlier than others. Therefore, this arrangement may not be the most memory efficient.

**Review by michaelbenayoun — APPROVED** [review]
